crates/proto/src/dnssec/handle.rs

```diff
@@ -1689,57 +1686,166 @@ fn verify_nsec(
                 )
-            }
+                && find_nsec_covering_record(soa_name, query.name(), nsecs).is_some() =>
+        {
+            nsec1_yield(
+                Proof::Secure,
+                query,
+                "no direct match, covering wildcard present for wildcard expansion response",
+            )
+        }
+        // For wildcard no data responses, we need to prove a wildcard matching wildcard_name does not contain
+        // the requested record type and that no closer match exists. (RFC 4035 3.1.3.4 and B.7/C.7)
+        None if !have_answer
+            && response_code == ResponseCode::NoError
+            && nsecs.iter().any(|(name, nsec_data)| {
+                name == &&wildcard_name
+                    && !nsec_data.type_set().contains(query.query_type())
+                    && !nsec_data.type_set().contains(RecordType::CNAME)
+                    && no_closer_matches(query.name(), soa_name, nsecs, wildcard_base_name.as_ref())
+            }) =>
+        {
+            nsec1_yield(
+                Proof::Secure,
+                query,
+                "no direct match, covering wildcard present",
+            )
         }
-        Some((_, _)) => nsec1_yield(
+        _ => nsec1_yield(
             Proof::Bogus,
             query,
-            format!("unsupported response code {response_code}"),
+            "no NSEC record matches or covers the wildcard name",
         ),
-        None => {
-            if !have_answer
-                && nsecs.iter().any(|(name, nsec_data)| {
-                    name == &&wildcard_name
-                        && !nsec_data.type_set().contains(query.query_type())
-                        && !nsec_data.type_set().contains(RecordType::CNAME)
-                })
-            {
-                return nsec1_yield(
-                    Proof::Secure,
-                    query,
-                    "no direct match, covering wildcard present",
-                );
-            }
+    }
+}
 
-            nsec1_yield(
-                Proof::Bogus,
-                query,
-                "no NSEC record matches or covers the wildcard name",
-            )
+/// Logs a debug message and returns a [`Proof`]. This is specific to NSEC validation.
+fn nsec1_yield(proof: Proof, query: &Query, msg: impl Display) -> Proof {
+    proof_log_yield(proof, query, "nsec1", msg)
+}
+
+// Prove that no closer name exists between the query name and wildcard_base_name
+fn no_closer_matches(
+    query_name: &Name,
+    soa: Option<&Name>,
+    nsecs: &[(&'_ Name, &'_ NSEC)],
+    wildcard_base_name: Option<&Name>,
+) -> bool {
+    let Some(wildcard_base_name) = wildcard_base_name else {
+        return false;
+    };
+
+    // If the SOA name is present, the query name and wildcard base name must be children of it.
+    if let Some(soa) = soa {
+        if !soa.zone_of(wildcard_base_name) {
+            debug!(%wildcard_base_name, %soa, "query_name is not a child of SOA");
```

Comment:
should this `debug!` say "widcard_base_name" instead of "query_name" in the message?